Web Inspector uses system highlight color, which becomes misleading when that's grey

AnujPanta1 opened the pull request:
#### e4ac810c5e5f4855566304c3c00c06dbed215cc9
<pre>
Web Inspector uses system highlight color, which becomes misleading when that&apos;s grey
<a href="https://bugs.webkit.org/show_bug.cgi?id=291237">https://bugs.webkit.org/show_bug.cgi?id=291237</a>
<a href="https://rdar.apple.com/148779029">rdar://148779029</a>

Reviewed by BJ Burg.

When macOS accent color is set to Graphite (gray), enabled toolbar buttons and selected DOM
elements become visually indistinguishable from disabled/unselected states. This patch detects
gray accent colors by checking if RGB channels differ by ≤ 20 (graphite is ~5, all chromatic
colors are &gt; 50) and overrides accent-derived CSS variables with blue. To handle changes while
Web Inspector is open, NSSystemColorsDidChangeNotification is observed in the UIProcess and
forwarded via IPC to WebInspectorUI, which dispatches systemAppearanceDidChange to
InspectorFrontendAPI to re-run the detection live.

* Source/WebInspectorUI/UserInterface/Base/Main.js:
* Source/WebInspectorUI/UserInterface/Protocol/InspectorFrontendAPI.js:
(InspectorFrontendAPI.systemAppearanceDidChange):
* Source/WebInspectorUI/UserInterface/Views/Variables.css:
(&amp;.mac-platform):
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::systemAppearanceDidChange):
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
(-[WKWebInspectorUIProxyObjCAdapter _systemColorsDidChange:]):
(WebKit::WebInspectorUIProxy::platformCreateFrontendPage):
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.cpp:
(WebKit::WebInspectorUI::systemAppearanceDidChange):
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.messages.in:

Canonical link: <a href="https://commits.webkit.org/307943@main">https://commits.webkit.org/307943@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f736d4d8c3461553434f9ffa09112a29429d65a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/146004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/18691 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/10844 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/154683 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/99531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/64cb7e96-40b3-48ad-b2f3-d9bba577a8cf) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/19170 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/18583 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/112315 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/99531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/1e30ec7e-6c1f-47a0-8bb9-522638481e07) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/148967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/14691 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/131152 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/93209 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/7dd2165d-2031-4fba-bd44-788742ca9ae8) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/13969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/11725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/2127 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/123526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/8095 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/156994 "Built successfully") | | 
| | [  ~~🛠 ios-safer-cpp~~](https://ews-build.webkit.org/#/builders/174/builds/206 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/9338 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/120328 "Passed tests") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/18522 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/15490 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/120655 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/30921 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/18553 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/129513 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/74231 "Built successfully") | | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/16356 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/7444 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/18141 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/81905 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/17877 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/18055 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/17937 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -1845,6 +1848,27 @@ WI._updateModifierKeys = function(event)
+        return;
+
+    let rgb = accentColor.rgb;
+    let channelDifference = Math.max(...rgb) - Math.min(...rgb);
```
> This patch detects gray accent colors by checking if RGB channels differ by ≤ 15 (graphite is ~5, all chromatic
colors are >50)

I'm unclear how this works. 15 _feels_ like a magic number.

If we're explicitly trying to detect gray, is `hsl` (Hue, *Saturation*, *Luminance*) better suited? (For whatever generally accepted / documented threshold "gray" has in HSL)

AnujPanta1 replied:
for me gray is usually defined by R G B being roughly equal, so what I really did with the <=15 is just making sure the spread between the channels is low, colorful colors will have a larger spread. This is also somewhat based on the idea that we know only graphite is the main issue. But I didn't know of a way to hard code it or something else

AnujPanta1 replied:
I didn't use HSL because getComputedStyle returns RGB, and I liked this better than a conversion because of it's size, but if you think this is too wishwashy I'd be okay with converting to HSL, might save us later if we ever allow accent colors to be customized

rcaliman-apple replied:
> I didn't use HSL because getComputedStyle returns RGB

You're getting RGB channels from `WI.Color.fromString()`:

```js
let accentColor = WI.Color.fromString(accentColorString);
if (!accentColor)
    return;

let rgb = accentColor.rgb; // you can ask for .hsl here
```

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/Variables.css:
```diff
@@ -434,3 +434,23 @@ body {
+        --selected-background-color: hsl(219, 80%, 43%);
+        --selected-text-background-color: hsl(230, 51%, 36%);
+    }
+}
```
Is there a potential CSS-only solution where [relative colors](https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Colors/Using_relative_colors) can be used to determine the saturation of the accent color and fallback to something else?

AnujPanta1 replied:
I don't think there is a pure CSS only solution, we still need JS for detection for the system color. Also, relative color more transforms colors I don't think that alone could correctly detect what color we need. Also I looked into `contrast-color` but not helpful in this situation

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/Variables.css:
```diff
@@ -434,3 +434,23 @@ body {
+
+body.mac-platform.low-contrast-accent .slider:focus {
+    border-color: hsl(212, 92%, 54%);
+}
```
At first I was confused where these colors come from, and then I realized they're the non-Mac platform colors.

---

I think we're bound to forget to update this over time if we use `-apple-system-accent-color` elsewhere.

How about assigning the accent color to a variable (instead of using it directly) that we override from JS instead of setting a class name on the body?

`Variables.css`
```
body {
  &.mac-platform {
--system-accent-color: -apple-system-accent-color;
  }
}
...

--breakpoint-color: var(--system-accent-color);
```

`Main.js`
```js
document.body.style.setProperty('--system-control-accent', "whatever blue accent color value is 😅")
```

This way, we remove this entire branch of `body.mac-platform.low-contrast-accent` styles.

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -1845,6 +1848,25 @@ WI._updateModifierKeys = function(event)
+        return;
+
+    let hsl = accentColor.hsl;
+    if (hsl[1] < 15)
```
Is this value sufficient to cover the case when Graphite is the system accent color?

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -1845,6 +1848,27 @@ WI._updateModifierKeys = function(event)
+
+    let hsl = accentColor.hsl;
+    if (hsl[1] < 20) {
+        let fallbackBlue = "hsl(212, 92%, 54%)";
```
```js
const fallbackBlue
```

burg commented on Source/WebKit/UIProcess/API/Cocoa/_WKInspector.mm:
```diff
@@ -215,6 +215,11 @@ - (void)_setDiagnosticLoggingDelegate:(id<_WKDiagnosticLoggingDelegate>)delegate
     protectedInspector(self)->setDiagnosticLoggingAvailable(!!delegate);
 }
 
+- (void)_systemAppearanceDidChange
```
Is this actually needed? We should get updates directly from the notification.

If there's a corresponding change for Safari please explain why needed. :)

burg commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -1845,6 +1848,39 @@ WI._updateModifierKeys = function(event)
+
+    let hsl = accentColor.hsl;
+    if (hsl[1] < 20) {
+        let isDark = window.matchMedia("(prefers-color-scheme: dark)").matches;
```
This is kinda against our style, BUT it does make it much easier to read compared to normal usage of spaces. Maybe a middle ground would be to put the data into two different objects and use object destructuring to fill these keys?

`let darkModeVariables = { fallbackAccent: ..., fallbackSelected: ..., ...}; lightModeVariables = { ... }; let {...} = isDark ? darkModeVariables : lightModeVariables`

burg reviewed (approved):
Looks good, just a question about the addition to WKInspector.h

AnujPanta1 committed:
Web Inspector uses system highlight color, which becomes misleading when that's grey
https://bugs.webkit.org/show_bug.cgi?id=291237
rdar://148779029

Reviewed by BJ Burg.

When macOS accent color is set to Graphite (gray), enabled toolbar buttons and selected DOM
elements become visually indistinguishable from disabled/unselected states. This patch detects
gray accent colors by checking if RGB channels differ by ≤ 20 (graphite is ~5, all chromatic
colors are > 50) and overrides accent-derived CSS variables with blue. To handle changes while
Web Inspector is open, NSSystemColorsDidChangeNotification is observed in the UIProcess and
forwarded via IPC to WebInspectorUI, which dispatches systemAppearanceDidChange to
InspectorFrontendAPI to re-run the detection live.

* Source/WebInspectorUI/UserInterface/Base/Main.js:
* Source/WebInspectorUI/UserInterface/Protocol/InspectorFrontendAPI.js:
(InspectorFrontendAPI.systemAppearanceDidChange):
* Source/WebInspectorUI/UserInterface/Views/Variables.css:
(&.mac-platform):
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::systemAppearanceDidChange):
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
(-[WKWebInspectorUIProxyObjCAdapter _systemColorsDidChange:]):
(WebKit::WebInspectorUIProxy::platformCreateFrontendPage):
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.cpp:
(WebKit::WebInspectorUI::systemAppearanceDidChange):
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorUI.messages.in:

Canonical link: https://commits.webkit.org/307943@main
